__tests__/integration.js

```diff
@@ -73,6 +73,62 @@ test('--mutex network', async () => {
   ]);
 });
 
+// Windows doesn't have the "echo" utility
+// Since this feature isn't platform-specific, running them on OSX and Linux only should be fine
+if (process.platform !== 'win32') {
+  test('yarn run <script> --opt', async () => {
+    const cwd = await makeTemp();
+
+    await fs.writeFile(
+      path.join(cwd, 'package.json'),
+      JSON.stringify({
+        scripts: {echo: 'echo'},
+      }),
+    );
+
+    const command = path.resolve(__dirname, '../bin/yarn');
+    const options = {cwd, env: {YARN_SILENT: 1}};
+
+    const {stderr: stderr, stdout: stdout} = execa(command, ['run', 'echo', '--opt'], options);
+
+    const stdoutPromise = misc.consumeStream(stdout);
+    const stderrPromise = misc.consumeStream(stderr);
+
+    const [stdoutOutput, stderrOutput] = await Promise.all([stdoutPromise, stderrPromise]);
```

Comment:
Problem with this writing is that it doesn't work, because the stdin and stdout streams are started and stopped as soon as the event loop gets a chance, ie. before the second line gets executed. Before of this, the second `consumeStream` call never receives the `end` event, and never returns. The synchronous version makes sure we attach both of our listeners before allowing the event loop to continue.